diff --git a/builtins.cpp b/builtins.cpp
--- a/builtins.cpp	2020-06-28 04:15:08.862852259 +0200
+++ b/builtins.cpp	2019-05-29 21:28:25.000000000 +0200
@@ -534,6 +534,16 @@
     return Error() << "Invalid code: " << code;
 }
 
+static Result<Success> do_install_keyring(const BuiltinArguments& args) {
+    if (e4crypt_install_keyring()) {
+        LOG(ERROR) << "failed to install keyring";
+        return Error() << "failed to install keyring";
+    }
+    property_set("ro.crypto.state", "encrypted");
+    property_set("ro.crypto.type", "file");
+    return Success();
+}
+
 /* mount_all <fstab> [ <path> ]* [--<options>]*
  *
  * This function might request a reboot, in which case it will
@@ -1050,6 +1060,7 @@
         {"init_user0",              {0,     0,    {false,  do_init_user0}}},
         {"insmod",                  {1,     kMax, {true,   do_insmod}}},
         {"installkey",              {1,     1,    {false,  do_installkey}}},
+        {"install_keyring",         {0,     0,    {false,  do_install_keyring}}},
         {"load_persist_props",      {0,     0,    {false,  do_load_persist_props}}},
         {"load_system_props",       {0,     0,    {false,  do_load_system_props}}},
         {"loglevel",                {1,     1,    {false,  do_loglevel}}},
